connect/handler.go

```diff
@@ -0,0 +1,329 @@
+	"github.com/inngest/inngest/pkg/syscode"
+	connectproto "github.com/inngest/inngest/proto/gen/connect/v1"
+	"github.com/inngest/inngestgo/internal/sdkrequest"
+	"github.com/pbnjay/memory"
+	"golang.org/x/sync/errgroup"
+	"io"
+	"log/slog"
+	"os"
+	"runtime"
+	"strings"
+	"sync"
+	"time"
+)
+
+const (
+	WorkerHeartbeatInterval = 10 * time.Second
+)
+
+func Connect(ctx context.Context, opts Opts, invoker FunctionInvoker, logger *slog.Logger) error {
+	ch := &connectHandler{
+		logger:  logger,
+		invoker: invoker,
+		opts:    opts,
+	}
+
+	wp := NewWorkerPool(ctx, opts.WorkerConcurrency, ch.processExecutorRequest)
+	ch.workerPool = wp
+
+	defer func() {
+		// TODO Push remaining messages to another destination for processing?
+	}()
+
+	err := ch.Connect(ctx)
+	if err != nil {
+		return fmt.Errorf("could not establish connection: %w", err)
+	}
+
+	return nil
+}
+
+type FunctionInvoker interface {
+	InvokeFunction(ctx context.Context, slug string, stepId *string, request sdkrequest.Request) (any, []state.GeneratorOpcode, error)
+}
+
+type Opts struct {
+	AppName string
+	Env     *string
+
+	Functions    []sdk.SDKFunction
+	Capabilities sdk.Capabilities
+
+	HashedSigningKey         []byte
+	HashedSigningKeyFallback []byte
+
+	WorkerConcurrency int
+
+	APIBaseUrl   string
+	IsDev        bool
+	DevServerUrl string
+	ConnectUrls  []string
+
+	InstanceId *string
+	BuildId    *string
+
+	Platform    *string
+	SDKVersion  string
+	SDKLanguage string
+}
+
+type connectHandler struct {
+	opts Opts
+
+	invoker FunctionInvoker
+
+	logger *slog.Logger
+
+	messageBuffer     []*connectproto.ConnectMessage
+	messageBufferLock sync.Mutex
+
+	hostsManager *hostsManager
+
+	workerPool *workerPool
+}
+
+// authContext is wrapper for information related to authentication
+type authContext struct {
+	hashedSigningKey []byte
+	fallback         bool
+}
+
+func (h *connectHandler) Connect(ctx context.Context) error {
+	signingKey := h.opts.HashedSigningKey
+	if len(signingKey) == 0 {
+		return fmt.Errorf("hashed signing key is required")
+	}
+
+	auth := authContext{hashedSigningKey: signingKey}
+
+	numCpuCores := runtime.NumCPU()
+	totalMem := memory.TotalMemory()
+
+	marshaledFns, err := json.Marshal(h.opts.Functions)
+	if err != nil {
+		return fmt.Errorf("failed to serialize connect config: %w", err)
+	}
+
+	marshaledCapabilities, err := json.Marshal(h.opts.Capabilities)
+	if err != nil {
+		return fmt.Errorf("failed to serialize connect config: %w", err)
+	}
+
+	hosts := h.connectURLs()
+	if len(hosts) == 0 {
+		return fmt.Errorf("no connect URLs provided")
+	}
+
+	h.hostsManager = newHostsManager(hosts)
+
+	// Notify when connect finishes (either with an error or because the context got canceled)
+	notifyConnectDoneChan := make(chan connectReport)
+
+	// Notify when connection is established
+	notifyConnectedChan := make(chan struct{})
+
+	// Channel to imperatively initiate a connection
+	initiateConnectionChan := make(chan struct{})
```

Comment:
why not make these channels part of the `connectHandler` instead of passing them around?

looks like it can be all done with a `NewConnectHandler` function.

Reply:
Sure could do that too